File: modules/ntsiniz-voice-dsp/ios/NtsinizVoiceDspModule.swift
import ExpoModulesCore
import Foundation

private struct DspSession {
  var noiseFloor: Double = 0.004
  var frames: Int = 0
  var voicedFrames: Int = 0
  var suppressionMode: String = "conservativeAdaptive"
}

public class NtsinizVoiceDspModule: Module {
  private var sessions: [String: DspSession] = [:]

  public func definition() -> ModuleDefinition {
    Name("NtsinizVoiceDsp")

    Function("createSession") { (opts: [String: Any]) -> String in
      return self.createSessionInternal(opts)
    }
    Function("processFrame") { (sessionId: String, pcmBase64: String, sampleRate: Int) -> [String: Any] in
      return try self.processFrameInternal(sessionId: sessionId, pcmBase64: pcmBase64, sampleRate: sampleRate)
    }
    Function("closeSession") { (sessionId: String) in
      self.closeSessionInternal(sessionId: sessionId)
    }

    AsyncFunction("createSessionAsync") { (opts: [String: Any]) -> String in
      return self.createSessionInternal(opts)
    }
    AsyncFunction("processFrameAsync") { (sessionId: String, pcmBase64: String, sampleRate: Int) -> [String: Any] in
      return try self.processFrameInternal(sessionId: sessionId, pcmBase64: pcmBase64, sampleRate: sampleRate)
    }
    AsyncFunction("closeSessionAsync") { (sessionId: String) in
      self.closeSessionInternal(sessionId: sessionId)
    }
  }

  private func createSessionInternal(_ opts: [String: Any]) -> String {
    let id = UUID().uuidString
    let suppression = (opts["suppressionMode"] as? String) ?? "conservativeAdaptive"
    self.sessions[id] = DspSession(noiseFloor: 0.004, frames: 0, voicedFrames: 0, suppressionMode: suppression)
    return id
  }

  private func closeSessionInternal(sessionId: String) {
    self.sessions.removeValue(forKey: sessionId)
  }

  private func processFrameInternal(sessionId: String, pcmBase64: String, sampleRate: Int) throws -> [String: Any] {
    guard var session = self.sessions[sessionId] else {
      throw NSError(domain: "NtsinizVoiceDsp", code: 1, userInfo: [NSLocalizedDescriptionKey: "session not found"])
    }
    guard let data = Data(base64Encoded: pcmBase64) else {
      throw NSError(domain: "NtsinizVoiceDsp", code: 2, userInfo: [NSLocalizedDescriptionKey: "bad base64"])
    }

    let bytesPerSample = 2
    let sampleCount = max(1, data.count / bytesPerSample)
    var sum = 0.0
    var peak = 0.0
    var crossings = 0
    var prev: Int16 = 0
    var hasPrev = false

    data.withUnsafeBytes { (raw: UnsafeRawBufferPointer) in
      let p = raw.bindMemory(to: Int16.self)
      for i in 0..<p.count {
        let s = p[i]
        let f = Double(s) / 32768.0
        sum += f * f
        let a = abs(f)
        if a > peak { peak = a }
        if hasPrev {
          if (prev < 0 && s > 0) || (prev > 0 && s < 0) {
            crossings += 1
          }
        } else {
          hasPrev = true
        }
        prev = s
      }
    }

    let rms = sqrt(sum / Double(sampleCount))
    let snrDb = 20.0 * log10((rms + 1e-7) / (session.noiseFloor + 1e-7))
    let zcr = Double(crossings) / Double(max(1, sampleCount - 1))

    let snrComponent = 1.0 / (1.0 + exp(-((snrDb - 6.0) / 4.5)))
    let levelComponent = 1.0 / (1.0 + exp(-((rms - 0.011) * 160.0)))
    let zcrPenalty = zcr > 0.35 ? 0.8 : 1.0
    let vadProb = min(1.0, max(0.0, (snrComponent * 0.62 + levelComponent * 0.38) * zcrPenalty))
    let voiced = vadProb >= 0.56 && rms >= session.noiseFloor * 1.12

    if !voiced {
      let smoothing = session.suppressionMode == "off" ? 0.02 : 0.08
      session.noiseFloor = min(0.18, max(0.0008, session.noiseFloor * (1.0 - smoothing) + rms * smoothing))
    } else {
      session.noiseFloor = min(0.18, max(0.0008, session.noiseFloor * 0.995 + rms * 0.005))
    }

    session.frames += 1
    if voiced { session.voicedFrames += 1 }
    self.sessions[sessionId] = session

    let voicedRatio = Double(session.voicedFrames) / Double(max(1, session.frames))
    let clipping = peak >= 0.985
    let signalQuality = qualityGrade(snrDb: snrDb, clipping: clipping, voicedRatio: voicedRatio)

    return [
      "vadProb": vadProb,
      "noiseFloorDb": 20.0 * log10(max(session.noiseFloor, 1e-7)),
      "snrDb": snrDb,
      "clipping": clipping,
      "voicedRatio": voicedRatio,
      "signalQuality": signalQuality,
      "sampleRate": sampleRate
    ]
  }
}

private func qualityGrade(snrDb: Double, clipping: Bool, voicedRatio: Double) -> String {
  if clipping { return "poor" }
  if snrDb < 5.0 { return "poor" }
  if snrDb < 10.0 { return "fair" }
  if voicedRatio < 0.12 { return "fair" }
  if snrDb < 17.0 { return "good" }
  return "excellent"
}

File: modules/ntsiniz-voice-dsp/android/src/main/java/com/ntsiniz/voicedsp/NtsinizVoiceDspModule.kt
package com.ntsiniz.voicedsp

import android.util.Base64
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap
import kotlin.math.abs
import kotlin.math.exp
import kotlin.math.log10
import kotlin.math.max
import kotlin.math.min
import kotlin.math.sqrt

private data class DspSession(
  var noiseFloor: Double = 0.004,
  var frames: Int = 0,
  var voicedFrames: Int = 0,
  var suppressionMode: String = "conservativeAdaptive"
)

class NtsinizVoiceDspModule : Module() {
  private val sessions = ConcurrentHashMap<String, DspSession>()

  override fun definition() = ModuleDefinition {
    Name("NtsinizVoiceDsp")

    Function("createSession") { opts: Map<String, Any?> ->
      createSessionInternal(opts)
    }
    Function("processFrame") { sessionId: String, pcmBase64: String, sampleRate: Int ->
      processFrameInternal(sessionId, pcmBase64, sampleRate)
    }
    Function("closeSession") { sessionId: String ->
      closeSessionInternal(sessionId)
    }

    AsyncFunction("createSessionAsync") { opts: Map<String, Any?> ->
      createSessionInternal(opts)
    }
    AsyncFunction("processFrameAsync") { sessionId: String, pcmBase64: String, sampleRate: Int ->
      processFrameInternal(sessionId, pcmBase64, sampleRate)
    }
    AsyncFunction("closeSessionAsync") { sessionId: String ->
      closeSessionInternal(sessionId)
    }
  }

  private fun createSessionInternal(opts: Map<String, Any?>): String {
    val id = UUID.randomUUID().toString()
    val suppression = (opts["suppressionMode"] as? String) ?: "conservativeAdaptive"
    sessions[id] = DspSession(suppressionMode = suppression)
    return id
  }

  private fun closeSessionInternal(sessionId: String) {
    sessions.remove(sessionId)
  }

  private fun processFrameInternal(sessionId: String, pcmBase64: String, sampleRate: Int): Map<String, Any> {
    val session = sessions[sessionId] ?: throw IllegalStateException("session not found")
    val bytes = try {
      Base64.decode(pcmBase64, Base64.DEFAULT)
    } catch (_: Throwable) {
      throw IllegalArgumentException("bad base64")
    }

    val sampleCount = max(1, bytes.size / 2)
    var sum = 0.0
    var peak = 0.0
    var crossings = 0
    var prev = 0.0
    var hasPrev = false

    var index = 0
    while (index + 1 < bytes.size) {
      val lo = bytes[index].toInt() and 0xFF
      val hi = bytes[index + 1].toInt() shl 8
      val sample = (hi or lo).toShort().toInt()
      val normalized = sample / 32768.0
      sum += normalized * normalized
      val absValue = abs(normalized)
      if (absValue > peak) peak = absValue
      if (hasPrev) {
        if (prev < 0 && normalized > 0 || prev > 0 && normalized < 0) crossings += 1
      } else {
        hasPrev = true
      }
      prev = normalized
      index += 2
    }

    val rms = sqrt(sum / sampleCount.toDouble())
    val snrDb = 20.0 * log10((rms + 1e-7) / (session.noiseFloor + 1e-7))
    val zcr = crossings.toDouble() / max(1, sampleCount - 1).toDouble()

    val snrComponent = 1.0 / (1.0 + exp(-((snrDb - 6.0) / 4.5)))
    val levelComponent = 1.0 / (1.0 + exp(-((rms - 0.011) * 160.0)))
    val zcrPenalty = if (zcr > 0.35) 0.8 else 1.0
    val vadProb = clamp01((snrComponent * 0.62 + levelComponent * 0.38) * zcrPenalty)
    val voiced = vadProb >= 0.56 && rms >= session.noiseFloor * 1.12

    if (!voiced) {
      val smoothing = if (session.suppressionMode == "off") 0.02 else 0.08
      session.noiseFloor = clampNoiseFloor(session.noiseFloor * (1.0 - smoothing) + rms * smoothing)
    } else {
      session.noiseFloor = clampNoiseFloor(session.noiseFloor * 0.995 + rms * 0.005)
    }

    session.frames += 1
    if (voiced) session.voicedFrames += 1
    sessions[sessionId] = session

    val voicedRatio = session.voicedFrames.toDouble() / max(1, session.frames).toDouble()
    val clipping = peak >= 0.985

    return mapOf(
      "vadProb" to vadProb,
      "noiseFloorDb" to (20.0 * log10(max(session.noiseFloor, 1e-7))),
      "snrDb" to snrDb,
      "clipping" to clipping,
      "voicedRatio" to voicedRatio,
      "signalQuality" to qualityGrade(snrDb = snrDb, clipping = clipping, voicedRatio = voicedRatio),
      "sampleRate" to sampleRate
    )
  }
}

private fun qualityGrade(snrDb: Double, clipping: Boolean, voicedRatio: Double): String {
  if (clipping) return "poor"
  if (snrDb < 5.0) return "poor"
  if (snrDb < 10.0) return "fair"
  if (voicedRatio < 0.12) return "fair"
  if (snrDb < 17.0) return "good"
  return "excellent"
}

private fun clamp01(v: Double): Double = min(1.0, max(0.0, v))

private fun clampNoiseFloor(v: Double): Double = min(0.18, max(0.0008, v))
